Move game interface to Arena

oliverlee opened the pull request.

oliverlee committed:
Move Player interface to Arena

Move player interaction (e.g. playing cards, buying cards, etc.) to the
Arena. This functionality has not yet been removed from Player. Add a
'big money' simulation to main to checking playing treasures and buying
cards.

oliverlee committed:
Add play action card to Arena

oliverlee committed:
Remove reference counted Supply from Player

This commit also removes functionality which has been duplicated in
Arena.

oliverlee committed:
Remove use of references for CardKind

oliverlee commented:
I want to re-enable tests before this is merged.

oliverlee committed:
Add tests for Player methods

oliverlee committed:
Re-enable tests for playing and buying cards

oliverlee committed:
Fix build warnings

mickvangelderen commented on dominion/src/dominion/arena.rs:
```diff
@@ -1,40 +1,515 @@
         }
     }
+
+    fn get_player(&self, player_id: usize) -> Result<&Player> {
```
`s/get_player/player/g`
consider adding
`current_player` and `current_player_mut` which use turn.player_id

oliverlee replied:
Done and done. However, I'm not sure how we will use `current_player` and `current_player_mut` right now.

mickvangelderen commented on dominion/src/dominion/arena.rs:
```diff
@@ -1,40 +1,515 @@
+        }
+    }
+
+    pub fn check_hand(&self, player_id: usize) -> Result<&CardVec> {
```
`s/check_hand/hand/g`

oliverlee replied:
done

mickvangelderen commented on dominion/src/dominion/arena.rs:
```diff
@@ -1,40 +1,515 @@
-    pub fn players(&mut self) -> &mut Vec<Player> {
-        &mut self.players
+    pub fn check_discard_pile(&self, player_id: usize) -> Result<&CardVec> {
+        Ok(&self.get_player(player_id)?.discard_pile)
```
```suggestion
        self.player(player_id).map(|player| &player.discard_pile)
```

oliverlee replied:
Done

mickvangelderen reviewed (changes requested):
If you don't expect people to pass an invalid player id, it is reasonable to panic in methods that accept it. You can then provide a `try_*` version of the function that returns a result. The `try_` function can be used if the input is not to be trusted and an invalid player id is to be expected.

get and check don't mean much to me and I have to agree with Kevlin Henney that they just make code harder to read.

you can make functions pub(crate) to only expose then to the crate to facilitate testing and accesing internals. This allows you to provide convenience function without having to expose them in your public api.

oliverlee commented:
> If you don't expect people to pass an invalid player id, it is reasonable to panic in methods that accept it. You can then provide a try_* version of the function that returns a result. The try_ function can be used if the input is not to be trusted and an invalid player id is to be expected.

Makes sense but I suggest we do that later as this may change a lot at the moment.

> get and check don't mean much to me and I have to agree with Kevlin Henney that they just make code harder to read.

I think this is a good idea. Thanks for point it out.

oliverlee committed:
Shorten names of Arena methods

Remove parts of Arena method names that can be inferred from the call.
Rename Player.played to Player.play_zone to better identify a location.
